File: Application/src/main/java/com/shareyourproxy/api/rx/command/AddUserGroupCommand.java
package com.shareyourproxy.api.rx.command;

import android.app.Service;
import android.os.Parcel;
import android.support.annotation.NonNull;

import com.shareyourproxy.api.domain.model.Group;
import com.shareyourproxy.api.domain.model.User;
import com.shareyourproxy.api.rx.RxBusDriver;
import com.shareyourproxy.api.rx.command.eventcallback.EventCallback;

import java.util.List;

import static com.shareyourproxy.api.rx.RxUserGroupSync.addUserGroup;

/**
 * Created by Evan on 6/8/15.
 */
public class AddUserGroupCommand extends BaseCommand {

    public static final Creator<AddUserGroupCommand> CREATOR = new Creator<AddUserGroupCommand>() {
        @Override
        public AddUserGroupCommand createFromParcel(Parcel in) {
            return new AddUserGroupCommand(in);
        }

        @Override
        public AddUserGroupCommand[] newArray(int size) {
            return new AddUserGroupCommand[size];
        }
    };
    private final static java.lang.ClassLoader CL = AddUserGroupCommand.class.getClassLoader();
    public final Group group;
    public final User user;

    /**
     * Public constructor.
     *
     * @param user  logged in user
     * @param group this events group
     */
    public AddUserGroupCommand(@NonNull User user, @NonNull Group group) {
        this.user = user;
        this.group = group;
    }

    private AddUserGroupCommand(Parcel in) {
        this((User) in.readValue(CL), (Group) in.readValue(CL));
    }

    @Override
    public EventCallback execute(Service service) {
        return addUserGroup(service, user, group);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeValue(user);
        dest.writeValue(group);
    }
}

File: Application/src/main/kotlin/com/shareyourproxy/api/rx/command/AddUserGroupCommand.kt
package com.shareyourproxy.api.rx.command

import android.content.Context
import android.os.Parcel
import android.os.Parcelable
import com.shareyourproxy.api.domain.model.Group
import com.shareyourproxy.api.domain.model.User
import com.shareyourproxy.api.rx.RxUserGroupSync.addUserGroup
import com.shareyourproxy.api.rx.command.eventcallback.EventCallback

/**
 * Add a new user group.
 * @param user  logged in user
 * @param group this events group
 */
internal final class AddUserGroupCommand(val user: User, val group: Group) : BaseCommand() {

    private constructor(parcel: Parcel) : this(parcel.readValue(CL) as User, parcel.readValue(CL) as Group)

    override fun execute(context: Context): EventCallback {
        return addUserGroup(context, user, group)
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeValue(user)
        dest.writeValue(group)
    }

    companion object {
        private val CL = AddUserGroupCommand::class.java.classLoader
        @JvmField val CREATOR: Parcelable.Creator<AddUserGroupCommand> = object : Parcelable.Creator<AddUserGroupCommand> {
            override fun createFromParcel(parcel: Parcel): AddUserGroupCommand {
                return AddUserGroupCommand(parcel)
            }

            override fun newArray(size: Int): Array<AddUserGroupCommand?> {
                return arrayOfNulls(size)
            }
        }
    }
}
